### ISSUE TITLE
NArray/Matrix as const throws compiler error

### ISSUE THREAD

**Issue by BenjaProg** [dash-project/dash#283]
Hey,
in my simplified example I want to call a function with a const refernce to a NArray/Matrix.
But then I get an compiler error.
~~~shell
main.cpp:27:30: error: passing ‘const local_type {aka const dash::LocalMatrixRef<unsigned int, 2, 2, dash::BlockPattern<2, (dash::MemArrange)1u, long int> >}’ as ‘this’ argument discards qualifiers [-fpermissive]
~~~
I made a very reduced example in [feat-283-constMatrix](https://github.com/dash-project/dash/blob/feat-283-constMatrix/dash/examples/ex.12.feat-283/main.cpp).
If the parameter in [line12](https://github.com/dash-project/dash/blob/feat-283-constMatrix/dash/examples/ex.12.feat-283/main.cpp#L12) is const then the compiler error arises.

This example is without really meaning, but to show the problem. 
The real example would be [cowichan_product](https://github.com/dash-project/dash-apps/blob/master/cowichan/product/main.cpp).
Everything is fine as long as NArray/Matrix is not a const parameter.

**Comment by BenjaProg** [comment]
I think that's just a missing feature in form of missing const versions of some function(s)?

**Comment by fmoessbauer** [comment]
I will have a look at this tonight. If you want to assist, please create an unit test which is not compilable (but should be). Then it's much easier to track down the problem and this also enhances the code coverage.

**Comment by BenjaProg** [comment]
@fmoessbauer you solved my example already, i just tested compiling the cowichan example and it worked fine 👍 in my branch feat-283-constMatrix is the example which should not compile without your changes.
I can write a test for matrix access but I can't do it in the next days, because I have to put more time into my bachelor thesis ;-)

**Comment by fmoessbauer** [comment]
I tested it with your example, so it should work. You can pull from the bug branch if necessary. Unfortunately there are still some things to discuss before we can merge the PR. And there are some problems regarding the CI API of github at the moment.

**Comment by fuchsto** [comment]
Fixed in PR #291

**Comment by BenjaProg** [comment]
With the latest changes my examples from above don't comile with const parameter anymore.
Trying to compile the example from the feat branch leads to following error.
[constMatrix compiler Error.txt](https://github.com/dash-project/dash/files/787597/constMatrix.compiler.Error.txt)
An Extract:

~~~c++
/home/dash/opt/dash-0.3.0/include/dash/matrix/internal/LocalMatrixRef-inl.h:442:21: error:
cannot convert 
‘dash::Matrix<      unsigned int, 2, long int, dash::BlockPattern<2, (dash::MemArrange)1u, long int> >* const’
to
‘dash::Matrix<const unsigned int, 2, long int, dash::BlockPattern<2, (dash::MemArrange)1u, long int> >*’
in assignment
   ref._refview._mat = _refview._mat;
                     ^
~~~

**Comment by fuchsto** [comment]
A container of const values doesn't seem useful as container elements would not be assignalbe.
It doesn't compile any more because DASH containers are const-correct since my changes in #291

**Comment by fmoessbauer** [comment]
No, I do not agree on that: Use cases like the following are common and we definitely should support them:

```c++
void foo(const dash::NArray<uint, 2> & matIn){
  // ...
  auto mPtr = matIn.local.row(0).lbegin();
  // ...
}
```

**Comment by fuchsto** [comment]
That is a const container, not a container of const.

**Comment by fmoessbauer** [comment]
Yes, but what is wrong with my sample code? This code does not compile, as `sub` modifies an internal state. Element accesses are of course another topic, but this problem is just not related to them.

**Comment by fuchsto** [comment]
That's unrelated to the example presented by @BenjaProg then.
Does your code compile with the following changes?

```c++
const uint * mPtr = matIn.local.row(0).lbegin();
```

Or:

```c++
const auto mPtr = ...
```

Note that `auto` does not state `const` explicitly. It would be resolved as `uint * const`, just asking out of curiosity.

**Comment by fuchsto** [comment]
... ah, it's in @BenjaProg 's code, I see:

https://github.com/dash-project/dash/blob/feat-283-constMatrix/dash/examples/ex.12.feat-283/main.cpp#L17

Yes, looks like this should actually work. Some conversion operator seems to be missing. Will report back.

**Comment by fuchsto** [comment]
@fmoessbauer `LocalMatrixRef.sub` does not modify an internal state. It operates on a copy of `LocalMatrixRef`. The issue was the assignment `Matrix<T, ...> * = Matrix<const T, ...> *` that can safely be `reinterpret_cast`ed.

@BenjaProg I pushed to your branch `feat-283-constMatrix` for testing. Seems to work now.

**Comment by fuchsto** [comment]
@BenjaProg @fmoessbauer Added the original use case of this issue as unit test and cleaned up example code in PR #296

**Comment by BenjaProg** [comment]
Nice work! Now everything works as intended!
Tested it with your test, the example and [cowichan product (with const as parameter [not pushed yet]) ](https://github.com/dash-project/dash-apps/blob/master/cowichan/product/main.cpp).
Thanks for adding the test by the way!

**Comment by BenjaProg** [comment]
There is another compiler error concerning `const matrices`.
In cowichan/randmat there's a print2D function which prints matrices to sdt::out.
The old signature looks like this: 
~~~C++
void print2d( T& mat )
~~~
and the new one like this:
~~~C++
void print2d( const T& mat )
~~~

with this i get the following compiler error:
~~~C++
randmat.cpp: In instantiation of ‘void print2d(const T&) [with T = dash::Matrix<unsigned char, 2, long int, dash::BlockPattern<2, (dash::MemArrange)1u, long int> >]’:
randmat.cpp:79:30:   required from here
randmat.cpp:12:34: error: no match for call to ‘(const dash::Matrix<unsigned char, 2, long int, dash::BlockPattern<2, (dash::MemArrange)1u, long int> >) (int&, int&)’
       cout << std::setw(3) << mat(i,j)<< " ";
                                  ^
In file included from /home/lehmannb/opt/dash-0.3.0/include/dash/Matrix.h:645:0,
                 from /home/lehmannb/opt/dash-0.3.0/include/dash/Container.h:72,
                 from /home/lehmannb/opt/dash-0.3.0/include/libdash.h:39,
                 from randmat.cpp:1:
/home/lehmannb/opt/dash-0.3.0/include/dash/matrix/internal/Matrix-inl.h:488:1: note: candidate: dash::Matrix<T, NumDimensions, IndexT, PatternT>::reference dash::Matrix<T, NumDimensions, IndexT, PatternT>::operator()(Args ...) [with Args = {int, int}; ElementT = unsigned char; int NumDimensions = 2; IndexT = long int; PatternT = dash::BlockPattern<2, (dash::MemArrange)1u, long int>; dash::Matrix<T, NumDimensions, IndexT, PatternT>::reference = dash::GlobRef<unsigned char>] <near match>
 Matrix<T, NumDim, IndexT, PatternT>
 ^
/home/lehmannb/opt/dash-0.3.0/include/dash/matrix/internal/Matrix-inl.h:488:1: note:   passing ‘const dash::Matrix<unsigned char, 2, long int, dash::BlockPattern<2, (dash::MemArrange)1u, long int> >*’ as ‘this’ argument discards qualifiers
make: *** [randmat] Error 1
~~~
[Here](https://github.com/dash-project/dash-apps/blob/6dc88bbf842a1484469cbd8dd993f542f39ac7ef/cowichan/randmat/randmat.cpp#L9) is the code i am talking about.
If you want to try it yourself, please be sure to use this commit i linked to.

**Comment by fuchsto** [comment]
That's supposed to fail:

https://github.com/dash-project/dash-apps/blob/6dc88bbf842a1484469cbd8dd993f542f39ac7ef/cowichan/randmat/randmat.cpp#L12

Should be `static_cast<typename T::value_type>(mat.at(i, j));`

More specific, return type of `const dash::Matrix<uint>::at` is `GlobRef<const uint>`, not `GlobRef<uint>`.

**Comment by BenjaProg** [comment]
Hm okay so i should have made the cast like this?:
~~~c++
static_cast<const uint>( mat(i,j) )
~~~
But i just checked it, completely without the cast and the error still remains!
By the way the cast is only needed when the type in the matrix is `uchar` so `cout` won't output garbage.

**Comment by fuchsto** [comment]
Indeed, some const accessors have been missing.
Fixed along with #322 in #341 with your use case added as test case.

**Comment by BenjaProg** [comment]
Thanks for the effort! 
But unfortunately there are still compiler errors.
Now they have changed a bit... i opened a dedicated branch for this bug at [dash-apps](https://github.com/dash-project/dash-apps/blob/bug-283-constMatrix/cowichan/randmat/randmat.cpp).
See current compiler errors in appendix.
[constMat(x,y).txt](https://github.com/dash-project/dash/files/867882/constMat.x.y.txt)

**Comment by fuerlinger** [comment]
Possibly related: The following doesn't compile any more when the array is passed as a const ref.

~~~
void find_min(const dash::Array<int>& arr) {
  auto xxx = dash::min_element(arr.begin(), arr.end());
}

int main(int argc, char* argv[])
{
  dash::init(&argc, &argv);
  
  dash::Array<int> arr(100);
  find_min(arr);
  
  dash::finalize();

  return EXIT_SUCCESS;
}
~~~

**Comment by fuchsto** [comment]
Aye, my test case uses `const Matrix<...>.at(i.j)`, not `const Matrix<...>.(i,j)`, I'm on it.
And `min_element` apparently is not const-correct and fails since global references / pointers / iterators are.

**Comment by fuchsto** [comment]
Fixed and unit-tested in #343

**Comment by BenjaProg** [comment]
thanks for the clarification!
